snforge_std/src/cheatcodes/events.cairo

```diff
@@ -29,50 +18,71 @@ struct Event {
     data: Array<felt252>
 }
 
-
-/// An event spy structure, along with the events collected so far in the test.
-/// `events` are mutable and can be updated with `fetch_events`.
+/// An event spy structure containing an offset marking from which event given Spy
+/// should fetch events.
 #[derive(Drop, Serde)]
 struct EventSpy {
-    _id: felt252,
-    events: Array<(ContractAddress, Event)>,
+    _event_offset: usize
+}
+
+/// A wrapper structure on an array of events to handle filtering smoothly.
+#[derive(Drop, Serde)]
+struct Events {
+    events: Array<(ContractAddress, Event)>
```

Comment:
Might possibly change that into `Array<Event>` and change `Event` into:
```
#[derive(Drop, Clone, Serde)]
struct Event {
    emitted_by: ContractAddress,
    keys: Array<felt252>,
    data: Array<felt252>
}
```
but then we'd have to change `assert_...` methods so that the interface would be the same (they would have to accept `events: @Array<Event>` instead of `events: @Array<(ContractAddress, T)>`). 

Then we'd have the problem with the fact that `assert_...` methods accept `T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>` - it doesn't ensure that the event we want to assert against would have the `emitted_by` field inside - we'd need another method on our `Event` struct like:
```
fn from<T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>>(starknet_event: T, contract_address: ContractAddress) -> Event {
    // conversion from starknet::Event to our Event
}
```

Reply:
Do you think there is an advantage to be gained?

Reply:
I kind of struggle to find one, it's strictly preferential, is it?

Reply:
There is also an option to not flatten the original event here, but rather nest it:

```
#[derive(Drop, Clone, Serde)]
struct EmittedEvent<T: starknet::Event<T>> {
    emitted_by: ContractAddress,
    event_data: T
}
```

Reply:
It is preferential, yes, though I find having one structure with all the data nicer than passing the tuple (`contract_address`, event) to `assert_...` functions